File: 735-asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();

        for(int i = 0; i < asteroids.length; i++) {
            if(stack.isEmpty() || stack.peek() < 0 || asteroids[i] > 0){
                stack.push(asteroids[i]);
            } else {
                int popped = 2000;
                while(!stack.isEmpty() && stack.peek() > 0) {
                    if(stack.peek() < -asteroids[i]) {
                        popped = stack.pop();
                    } else break;
                }
                
                if(!stack.isEmpty() && stack.peek() == -asteroids[i]) {
                    stack.pop();
                } else if (stack.isEmpty() || (!stack.isEmpty() && stack.peek() < -asteroids[i])) {
                    stack.push(asteroids[i]);
                }
            }
        }

        int[] ans = new int[stack.size()];
        int i = 0;
        for(int element : stack) {
            ans[i] = element;
            i++;
        }

        return ans;
    }
}

File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for ast in asteroids:
            if len(stack) == 0:
                stack.append(ast)
                continue

            if stack[-1] > 0 and ast < 0:
                while len(stack) > 0 and stack[-1] > 0 and stack[-1] < abs(ast):
                    stack.pop()
                if len(stack) > 0 and stack[-1] == abs(ast):
                    stack.pop()
                elif len(stack) == 0 or stack[-1] < 0:
                    stack.append(ast)
            else:
                stack.append(ast)

        return stack